packages/web-api-scan-runner/src/crawler/discovered-urls-processor.ts

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { ServiceConfiguration } from 'common';
+import { inject, injectable } from 'inversify';
+import _, { isNil } from 'lodash';
+import { WebsiteScanResult } from 'storage-documents';
+
+@injectable()
+export class DiscoveredUrlsProcessor {
+    public constructor(@inject(ServiceConfiguration) private readonly serviceConfig: ServiceConfiguration) {}
+
+    public async getProcessedUrls(discoveredUrls: string[], websiteScanResult?: WebsiteScanResult): Promise<string[]> {
+        let processedUrls = discoveredUrls;
+        if (isNil(discoveredUrls)) {
+            return [];
+        }
+
+        if (!isNil(websiteScanResult) && !isNil(websiteScanResult.knownPages)) {
+            processedUrls = this.removeUrlsFromList(discoveredUrls, websiteScanResult.knownPages);
+        }
+        processedUrls = this.limitNumUrls(processedUrls, await this.getUrlLimit(websiteScanResult));
+
+        return processedUrls;
+    }
+
+    private removeUrlsFromList(urlList: string[], removeUrls: string[]): string[] {
+        return _.pullAll(urlList, removeUrls);
+    }
+
+    private limitNumUrls(urlList: string[], urlLimit: number): string[] {
+        return _.take(urlList, urlLimit);
+    }
+
+    private async getUrlLimit(websiteScanResult?: WebsiteScanResult): Promise<number> {
+        const urlCrawlLimit = (await this.serviceConfig.getConfigValue('crawlConfig')).urlCrawlLimit;
```

Comment:
Is it ever possible for the crawl limit to be a negative number?

Reply:
We configure the crawl limit ourselves, so while it's possible for it to be set to a negative number, in practice that's not something we'd ever expect to see